### PR TITLE
feat(dspy): add DSPy instrumentation

### LINKED ISSUE #1166: 🚀 Feature: Support DSPy
### Which component is this feature for?

All Packages

### 🔖 Feature description

Instrument [DSPy](https://github.com/stanfordnlp/dspy) framework

### 🎤 Why is this feature needed ?

-

### ✌️ How do you aim to achieve this?

Similar to our LlamaIndex & LangChain instrumentations

### 🔄️ Additional Information

_No response_

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### Are you willing to submit PR?

None

### PR DESCRIPTION
**Pull request by banlor**
Closes #1166.

picked this up since houseman313's december plan didn't turn out with a PR.

wraps at the LM boundary (`LM.forward` / `aforward`) rather than `Evaluate` — same conclusion someshfengde reached in september: wrapping evaluate alone misses calls during compile/bootstrap. crewai felt like the right reference here since dspy doesn't expose callbacks the way langchain does.

**what's instrumented**

- `LM.forward` / `aforward` → client span per LM call. captures model, provider, input/output messages (incl. tool calls), token usage, duration, `dspy.cache_hit`
- `Predict.forward` / `aforward` → internal task span, parent of the LM span. captures signature name via `dspy.signature`

provider name uses `GenAISystem` / `GenAiSystemValues` enums straight from semconv-ai and the otel upstream — same pattern as the groq alignment in 4c6d2743. covers litellm prefixes (`openai/`, `anthropic/`, `bedrock/`, `azure/`, `vertex_ai/`, `gemini/`, `groq/`, `mistral/`, `cohere/`, `ollama/`) plus bare model patterns.

**tested**

ran locally against `anthropic/claude-haiku-4-5-20251001`, both spans come through with the right attributes. 21 unit tests cover sync + async paths, exception handling (input messages survive failures), predict→lm parent-child, tool-call output, cache-hit zero-token edge case, and parametrized provider inference across all supported prefixes. mocks the litellm layer so they run without api keys. cassettes can be added in a follow-up if you prefer.

also added `dspy_example.py` to the sample-app package.

## Summary by CodeRabbit

* **New Features**
  * OpenTelemetry DSPy instrumentation: traces LM and Predict calls with GenAI spans, inferred provider/model attributes, token-usage & duration metrics, cache-hit reporting, and parent/child linkage. Metrics can be disabled via env var.

* **Tests**
  * Expanded coverage for LM/ Predict spans, provider inference, token counts, cache behavior, tool-call serialization, async flows, parent/child linkage, and re-instrumentation.

* **Documentation**
  * New README with install, usage, emitted spans/attributes, and metrics notes.

* **Examples**
  * Added a runnable DSPy example demonstrating instrumentation.

* **Packaging**
  * Package metadata and instrumentor entry point registered.

[Review Change Stack](https://app.coderabbit.ai/change-stack/traceloop/openllmetry/pull/4089)

### CODE REVIEW

File: packages/opentelemetry-instrumentation-dspy/opentelemetry/instrumentation/dspy/instrumentation.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+    provider = _infer_provider(model)
+
+    span_attrs = {
+        GenAIAttributes.GEN_AI_OPERATION_NAME: GenAiOperationNameValues.CHAT.value,
+        GenAIAttributes.GEN_AI_REQUEST_MODEL: model,
+    }
+    if provider:
+        span_attrs[GenAIAttributes.GEN_AI_PROVIDER_NAME] = provider
+
+    with tracer.start_as_current_span(model, kind=SpanKind.CLIENT, attributes=span_attrs) as span:
```
[the remark this answers is not in the document]
Comment (banlor) [reply packages/opentelemetry-instrumentation-dspy/opentelemetry/instrumentation/dspy/instrumentation.py]: fixed

File: packages/opentelemetry-instrumentation-dspy/opentelemetry/instrumentation/dspy/instrumentation.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+            )
+            result = wrapped(*args, **kwargs)
+
+            _set_lm_span_output(span, messages, result, instance, model, provider,
+                                 duration_histogram, token_histogram, start)
+            span.set_status(Status(StatusCode.OK))
+            return result
+        except Exception as ex:
+            span.set_status(Status(StatusCode.ERROR, str(ex)))
+            raise
```
[the remark this answers is not in the document]
Comment (banlor) [reply packages/opentelemetry-instrumentation-dspy/opentelemetry/instrumentation/dspy/instrumentation.py:189]: done

File: packages/opentelemetry-instrumentation-dspy/project.json [hunk 3]
```diff
@@ -0,0 +1,77 @@
+{
+  "name": "opentelemetry-instrumentation-dspy",
+  "$schema": "../../node_modules/nx/schemas/project-schema.json",
+  "projectType": "library",
+  "sourceRoot": "packages/opentelemetry-instrumentation-dspy/opentelemetry_instrumentation_dspy",
```
[the remark this answers is not in the document]
Comment (banlor) [reply packages/opentelemetry-instrumentation-dspy/project.json]: fixed (ef55fd18).

File: packages/opentelemetry-instrumentation-dspy/pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  "pytest>=8.2.2,<9",
+  "ruff>=0.4.0",
+]
+test = [
+  "dspy-ai>=2.0.0",
+  "opentelemetry-sdk>=1.38.0,<2",
+  "pytest-recording>=0.13.1,<0.14.0",
+  "pytest-sugar==1.0.0",
+  "pytest>=8.2.2,<9",
+]
```
[the remark this answers is not in the document]
Comment (banlor) [reply packages/opentelemetry-instrumentation-dspy/pyproject.toml:41]: done

File: packages/sample-app/sample_app/dspy_example.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import dspy
+from traceloop.sdk import Traceloop
+
+Traceloop.init(app_name="dspy-example")
+
+lm = dspy.LM("openai/gpt-4o-mini")
+dspy.configure(lm=lm)
```
[the remark this answers is not in the document]
Comment (banlor) [reply packages/sample-app/sample_app/dspy_example.py]: fixed

File: packages/opentelemetry-instrumentation-dspy/tests/test_dspy_instrumentation.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+    assert len(spans) == 1
+    span = spans[0]
+    assert span.attributes[GenAIAttributes.GEN_AI_OPERATION_NAME] == "chat"
+    assert span.attributes[GenAIAttributes.GEN_AI_USAGE_INPUT_TOKENS] == 10
+
+
+def test_predict_forward_span_name_and_no_chat_op(span_exporter):
+    lm = dspy.LM("openai/gpt-4o", cache=False)
+    dspy.configure(lm=lm)
+    predict = dspy.Predict("question -> answer")
```
[the remark this answers is not in the document]
Comment (banlor) [reply packages/opentelemetry-instrumentation-dspy/tests/test_dspy_instrumentation.py]: fixed

**Comment by banlor** [comment]
@coderabbitai resume

File: packages/opentelemetry-instrumentation-dspy/opentelemetry/instrumentation/dspy/instrumentation.py [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+
+        start = time.time()
+        try:
+            result = wrapped(*args, **kwargs)
+        except Exception as ex:
+            span.set_status(Status(StatusCode.ERROR, str(ex)))
+            raise
+        _safe_set_lm_span_output(span, result, model, provider, duration_histogram, token_histogram, start)
+        span.set_status(Status(StatusCode.OK))
+        return result
```
[the remark this answers is not in the document]
Comment (banlor) [reply packages/opentelemetry-instrumentation-dspy/opentelemetry/instrumentation/dspy/instrumentation.py:192]: Skipping. New event is still Development status; rest of the repo uses `record_exception()`. Repo-wide switch would be a separate PR.

**Comment by banlor** [comment]
@galkleinman could you take a look at the DSPy changes here and the fixes in #4104 and #4459?
